### ISSUE TITLE
Celestial_LocalTime is very slow

### ISSUE THREAD

**Issue by dm413** [Tronald/CoordinateSharp#94]
I am working on a project that needs to know sunset and sunrise in local time for a large number of dates. It is running very slow, and I tracked this down to Celestial_LocalTime, which takes 35 milliseconds per call. Everything else in the library runs in a few microseconds.

Is this a bug? Is there anything I can do to speed it up? Thanks for any advice/help.

**Comment by Tronald** [comment]
Hi there, thanks for checking out CoordinateSharp.

So the slowness you are experiencing is partially expected. The reason you see this when converting to local time is because the calculations that are occurring have to run 3 times (day prior, day of, and next day). Once these calculations are completed, then the logic within the library determines the next event at the local time. This has to be done as the event day could change when operating in local time.

This delay normally goes unnoticed until you do batch conversions (like you are doing).

Celestial calculations take approximately 8ms to complete. So it should take approximately 24ms for local time conversion (8 x 3=24). The other 9ms is most likely coming from your coordinate constructors. You are probably calling a new Coordinate in each loop. We may be able to modify your logic to increase performance.

Can you post a code example of how you are iterating and converting through the coordinates? We may be able to shave of some milliseconds.

**Comment by dm413** [comment]
Inside the loop I am doing:
```
DateTime dtutc = dt.ToUniversalTime();      // dt is the date/time in local time
Coordinate coordinate = new Coordinate(Latitude, Longitude, dtutc);
Celestial celestial = Celestial.Celestial_LocalTime(coordinate, timeZoneOffset);
```
And you're right, the Coordinate constructor takes about 10 ms. The call to Celestial_LocalTime takes 35 ms just by itself. The loop time is about 46 ms.

By the way, is there any way to get a timeZone offset from a coordinate?

Thanks,

**Comment by Tronald** [comment]
Great thank you!

You sacrifice 10 ms as you said with each Coordinate creation. By turning eager loading off you will shave 9.9999 ms off your benchmark. You don't need to eager load the `Coordinate` values as you simply need the `Coordinate` object to pass to the local time conversion. This option will improve performance by about 25% and keep your accuracy. It's not perfect, but it's something.

Change

```C#
Coordinate coordinate = new Coordinate(Latitude, Longitude, dtutc);
```

to 

```C#
Coordinate coordinate = new Coordinate(Latitude, Longitude, dtutc, new EagerLoad(false));
```

My benchmark went from 32 ms to 23 ms (this will obviously be based on CPU).

I plan to review this feature prior to releasing **v2.1.1.1**. When celestial calculations convert to local time it converts everything. I want to add options to just convert solar or lunar times as it could save a lot of time for situations like yours. I also plan to eventually rework the trigonometry involved in order to do direct timezone calculations (which would shave it down to 8 ms), but that is a very complex process and I just don't have the time to dedicate at the moment.

Regarding your question about getting the timezone offset, there is no way to do it within the library, but I do know how to get it. It's a doozy though.

Your best bet is to query Google for a IANA timezone ID, and then get the offset using [NodaTime](https://www.nuget.org/packages/NodaTime).

Check out this [StackOverflow](https://stackoverflow.com/questions/33639571/get-local-time-based-on-coordinates) post on how to query Google for a timezone. The field you want to concentrate on getting is `timeZoneID`.

Once you have pulled the info from Google, you can get the offset using NodaTime [StackOverflow Post](https://stackoverflow.com/questions/40006629/get-offset-minutes-from-timezone-string-with-nodatime).

This process is SLOWWWW, and Google limits your requests per day. I had to a batch grab just once and it sucked, but it's really the only free way to do this (or at least was a few years ago when I had to do something similar).

Let me know how the performance increase is with eager loading turned off. I am not sure how soon you need to accomplish your task, but if you are still struggling I can send you a testing copy of v2.1.1.1 once the additional local time features are added to assist you. 

-Justin

**Comment by dm413** [comment]
Thanks. Adding EagerLoad(false) did eliminate the time for the constructor which helps. I also added a dictionary to cache the sunrise and sunset times, which got me about a 4x improvement. (I am processing photographs, so the same date occurs more than once.) It's still slower than I would like, but it's acceptable now. 

Thanks for your quick and thorough help with this.

**Comment by Tronald** [comment]
No problem. Good call with the dictionary! Thanks for letting me know.

**Comment by Tronald** [comment]
Good Afternoon.

I'm reaching out to let you know that we will be adding at Solar and Lunar only Local Time converter. This will allow you to generate a local time `Celestial` objects with only the solar or Lunar times populated. This will reduce the benchmark from 24ms to 7ms since you are only needing to acquire sunrise/sunset times.

I expect this will increase performance dramatically.

It will be available when v2.1.1.1 release (hopefully here in the next few weeks).

**Comment by Tronald** [comment]
Good Morning.

I am updating this issue to let you know a v2.1.1.1 pre-release is out (here and on Nuget) that can better handle this subject by allowing for local suntime calculations only. This will cut the benchmark by more then half.

```C#
Coordinate coordinate = new Coordinate(45, 67, DateTime.Now, new EagerLoad(false));
Celestial celestial = Celestial.Solar_LocalTime(coordinate, -5);
Console.WriteLine(celestial.SunSet);
```

**Comment by Tronald** [comment]
Good Evening, I hope all is well.

The trig has been reworked for local time conversions. Benchmarks will now match UTC time benchmarks.

Celestial Calcs - 5ms
Solar Only - 2ms
Lunar Only  3ms

These improvements will be made available in the next release (2.3.1.1). We are currently testing the new formulas, but everything is looking good so far to get them out by next release. This should greatly help.
